bolt/lib/Core/BinaryFunction.cpp

```diff
@@ -1673,12 +1678,104 @@ bool BinaryFunction::scanExternalRefs() {
     if (BranchTargetSymbol) {
       BC.MIB->replaceBranchTarget(Instruction, BranchTargetSymbol,
                                   Emitter.LocalCtx.get());
-    } else if (!llvm::any_of(Instruction,
-                             [](const MCOperand &Op) { return Op.isExpr(); })) {
-      // Skip assembly if the instruction may not have any symbolic operands.
-      continue;
     } else {
+      bool NeedsPatch = false;
       analyzeInstructionForFuncReference(Instruction);
+      for (unsigned OpNum = 0; OpNum < MCPlus::getNumPrimeOperands(Instruction);
+           ++OpNum) {
+        const MCSymbol *Symbol = BC.MIB->getTargetSymbol(Instruction, OpNum);
+        if (!Symbol)
+          continue;
+        if (!ignoreReference(Symbol)) {
+          NeedsPatch = true;
+          break;
+        }
+      }
+      if (!NeedsPatch)
+        continue;
+    }
+
+    // For AArch64, we need to undo relaxation done by the linker if the target
+    // of the instruction is a function that we plan to move.
+    const Relocation *Rel;
+    if (BC.isAArch64() && (Rel = getRelocationAt(Offset))) {
+      // NOP+ADR sequence can originate from either ADRP+ADD or ADRP+LDR.
+      // In either case, we convert it into ADRP+ADD.
+      if (BC.MIB->isADR(Instruction) &&
+          (Rel->Type == ELF::R_AARCH64_ADD_ABS_LO12_NC ||
+           Rel->Type == ELF::R_AARCH64_LD64_GOT_LO12_NC)) {
+        if (!BC.MIB->isNoop(PrevInstruction)) {
+          const MCSymbol *Symbol = BC.MIB->getTargetSymbol(Instruction);
+          BC.errs() << "BOLT-WARNING: cannot undo linker relaxation for "
+                       "instruction at 0x"
+                    << Twine::utohexstr(AbsoluteInstrAddr) << " referencing "
+                    << Symbol->getName() << '\n';
+          if (BinaryFunction *TargetBF = BC.getFunctionForSymbol(Symbol))
+            TargetBF->setIgnored();
+          continue;
```

Comment:
This is ignored for similar reasons as with [line 1730](https://github.com/llvm/llvm-project/pull/133014/files#diff-ca39122670f58da421af7031c33576d77078d2b81f48aa9baccd945724edb891R1730), right?

We have the correct relocations here but the nop is missing, so it's better to be safe.

nit: consider adding a comment

Reply:
At 1730, it's a code from compiler/assembler without the linker involvement. I'll add a comment.